src/bio/bio_wal.c

```diff
@@ -449,6 +449,7 @@ place_payload(struct bio_sglist *bsgl, struct wal_blks_desc *bd, struct wal_tran
 
 		copy_sz = (left >= len) ? len : left;
 		memcpy(tb->tb_buf + tb->tb_off, (void *)addr, copy_sz);
+		VALGRIND_MAKE_MEM_DEFINED(tb->tb_buf + tb->tb_off, copy_sz);
```

Comment:
The WAL payload isn't only for DAV snapshots, I don't think it's correct to mark all 'payload' as defined.

Even if it's only for DAV snapshots, I'm not quite sure why we need to mark it as defined here?

Reply:
@NiuYawei the payload can end up having address ranges that are potentially marked undefined. This is due to the requirement that every allocated memory block should be persisted at the end of a transaction. This is not an issue with PMDK and Valgrind. However with DAV, persist implies creating a redo log. This results in valgrind setting some of the memory ranges in the payload as undefined. This payload when copied to DMA buffer will propagate it there too. Later, when this DMA buffer is used for some other read it is wrongly propagating these undefined memory ranges.

We tried to fix this in DAV while creating the payload in dav redo log. However, this will not work with UMEM_ACT_COPY_PTR which we use for dav_reserve(). Hence placed the change at the above location.

@ivadmitry I believe macro should be called only when DAOS_ON_VALGRIND is defined.

Reply:
Please see the updated patch.